[features] Media Library list view

ronronscelestes opened the pull request:
## What

The media library display is evolving offering the possibility to switch between a list view and a grid view.
The goal is to give an overview of the asset details to have maximum information available at first glance.
This is applied in the Media Library and the Asset Dialog.

While working on the list view, some fixes were also brought to the permissions grid view as users without update permission should not be able to access move or delete actions.

## Snapshots

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/71838159/205687023-b786c6fb-cf98-4bb5-b4d6-e00af325a014.png">

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/71838159/205687092-0f68b9af-1963-4307-bd99-8a84f1e1555b.png">


## Tests

Switch from list view to grid view in ML and ML in CM
Check that all permissions are working the same for grid and list view

### Acceptance criteria

- Given that I am in the Media Library
  - When I click on the List view icon
    - I cannot see the Sort by button
    - I can sort on `Created`, `Last updated`, `Name` by clicking table headers arrow icon  
  - When I click on a folder row
    - I’m redirected to the folder page
  - When I click on an asset row 
    - I can see the edit asset modal open
  - When I don’t have update permissions
    - I can still access edit view of a folder or an asset
    - I can see a disabled form (inputs + actions buttons)
    - I cannot select any folder or asset
- Given than I am in the Media Library through the Media Library Input in the CM (modals)
  - All the above is true 
    - Except **When I don’t have update permissions → I cannot select any folder or asset**
  - When I want to select
    - I cannot select folders
    - I can only select allowed assets (through field advanced settings in CTB)

joshuaellis commented on packages/core/upload/admin/src/components/AssetDialog/BrowseStep/utils/isSelectable.js:
```diff
@@ -0,0 +1,18 @@
+/**
+ * @typedef isSelectable
```
This isn't doing what you expect it too:

```js
/**
 * @param {string[]} allowedTypes
 * @param {string} mime
 * @returns {boolean}
 */
```

joshuaellis commented on packages/core/upload/admin/src/utils/getFolderURL.js:
```diff
@@ -1,11 +1,11 @@
 import { stringify } from 'qs';
 
-const getFolderURL = (pathname, query, folder) => {
+const getFolderURL = (pathname, query, folderID) => {
```
worth adding a JSdoc & clarifying that `folderID` can be `undefined` and thats okay?

joshuaellis reviewed (changes requested):
Really great work @ronronscelestes & @jhoward1994 ⭐ 

Some minor feedback just around comments / jsdocs 😄

ronronscelestes commented on packages/core/upload/admin/src/components/AssetDialog/BrowseStep/index.js:
```diff
@@ -195,99 +240,135 @@ export const BrowseStep = ({
+          onSelectOne={onSelectAsset}
+          onSelectAll={onSelectAllAsset}
+          rows={
+            // TODO: remove when fixed on DS side
```
indeed I'll remove this workaround and test that it's properly working

jhoward1994 reviewed (dismissed):
Great work! 🎉 it works really well

Marc-Roig commented:
QA done & LGTM! Amazing work 🚀

gu-stav commented on packages/core/upload/admin/src/components/TableList/CellContent.js:
```diff
@@ -0,0 +1,72 @@
+    case 'size':
+      if (elementType === 'folder') return <Typography>-</Typography>;
+
+      return <Typography>{formatBytes(content)}</Typography>;
```
Are we displaying only the date or datetime here? I'm wondering if it would make sense to unify this with the behavior in the CM: https://github.com/strapi/strapi/blob/45831f30f5993709fa0fe06222578603d0cc4eb9/packages/core/admin/admin/src/content-manager/components/DynamicTable/CellContent/CellValue.js#L10-L16

ronronscelestes replied:
[fixed](https://github.com/strapi/strapi/pull/15085/commits/dc055b43710f3606b5f450f318f99638d50ae00e)

gu-stav replied:
Sounds good to me, thanks!

gu-stav commented on packages/core/upload/admin/src/components/TableList/CellContent.js:
```diff
@@ -0,0 +1,72 @@
+      return <Typography>{content}</Typography>;
+
+    default:
+      return <Typography>-</Typography>;
```
Just as a suggestion: I'm not sure if this is a good a11y practice (same for the CM) and I'm wondering if it would be better if we would render an aria-label here saying, this field is empty or defining aria-hidden on the value?

Opinions, @joshuaellis ?

ronronscelestes replied:
I like the idea of an aria-label!

ronronscelestes replied:
added a translated aria label [here](https://github.com/strapi/strapi/pull/15085/commits/dc055b43710f3606b5f450f318f99638d50ae00e)

joshuaellis replied:
i like the label.

gu-stav commented on packages/core/upload/admin/src/components/TableList/PreviewCell.js:
```diff
@@ -0,0 +1,69 @@
+    );
+  }
+
+  if (mime.includes('image')) {
+    const mediaURL = prefixFileUrlWithBackendUrl(thumbnailURL) ?? prefixFileUrlWithBackendUrl(url);
+
+    return <Avatar src={mediaURL} alt={alternativeText} preview />;
+  }
```
Just wondering: in the grid view we render preview images for videos. Does that mean in the list view we don't do that?

ronronscelestes replied:
good question - it wasn't discussed 🤔
since `Avatar` could not handle it right now (we would need a new component), wdyt of discussing it quickly with product and iterate on it after the release?

gu-stav replied:
Sounds good to me. Can you create a follow-up issue on Jira? 🚀

ronronscelestes replied:
CONTENT-828 created 🚀

gu-stav commented on packages/core/upload/admin/src/pages/App/MediaLibrary.js:
```diff
@@ -255,11 +300,38 @@ export const MediaLibrary = () => {
+              onSelectOne={selectOne}
+              onSelectAll={handleBulkSelect}
+              rows={
+                // TODO: remove when fixed on DS side
+                // when number of rows in Table changes, the keyboard tab from a row to another
+                // is not working for 1st and last column
+                !assetsLoading && !foldersLoading ? [...folders, ...assets] : []
```
Do we have an issue for that in the DS?

ronronscelestes replied:
my bad I thought I removed that already - fixed by Josh in the DS!

ronronscelestes replied:
[fixed](https://github.com/strapi/strapi/pull/15085/commits/dc055b43710f3606b5f450f318f99638d50ae00e)

gu-stav reviewed (commented):
Well done! 🚀 Added only a few minor questions ...

gu-stav commented on packages/core/upload/admin/src/translations/en.json:
```diff
@@ -115,14 +115,15 @@
-  "list-table-header-createdAt": "created",
-  "list-table-header-updatedAt": "last update",
-  "list-table-header-sort": "Sort on {label}",
+  "list.table.header.actions": "actions",
```
Oh thanks. I thought the same but didn't want to start bikeshedding :D

ronronscelestes replied:
I don't understand why I used this syntax, triggered me when I saw it 😂

ronronscelestes commented:
sorry for the approval dismissal, I updated the branch on main + caught a small UI issue (e.g. removed the table in `BrowseStep` when no assets or folders, and added a test case for it too)
@gu-stav @joshuaellis @jhoward1994 @Marc-Roig

jhoward1994 reviewed (approved):
LGTM 👌🏻
